Initial attempt at implementing property-based testing using Hypothesis

RohitP2005 opened the pull request:
# Description

This change introduces the use of Hypothesis-based testing in two files and is being applied to selective functions as requested by the author and members of the issue https://github.com/pybamm-team/PyBaMM/issues/4703#issuecomment-2564062105

Starts #4703

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [x] Optimization (back-end change that speeds up the code)
- [X] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python -m pytest` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python -m pytest --doctest-plus src` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

prady0t commented:
@RohitP2005 Add `hypothesis` as a dependency in `pyproject.toml` for checks to pass. 

https://github.com/pybamm-team/PyBaMM/blob/421784e6b0a5f0e857be155b693fec1d9e7ba72d/pyproject.toml#L100

prady0t commented on tests/unit/test_expression_tree/test_averages.py:
```diff
@@ -5,29 +5,32 @@
             pybamm.PrimaryBroadcast(a, ["negative electrode"])
         )
-        assert average_broad_a == pybamm.Scalar(4)
+        assert average_broad_a == pybamm.Scalar(random_value)
```
```suggestion
        assert average_broad_a == pybamm.Scalar(random_integer)
```

prady0t reviewed (commented):
I've added reviews for `test_averages.py`. I'll add reviews for `test_unary_operators.py` once the failing tests are fixed.

prady0t commented:
Also, fix the PR heading. It's not proper.

RohitP2005 commented on tests/unit/test_expression_tree/test_averages.py:
```diff
@@ -174,9 +177,10 @@ def test_x_average(self):
         assert pybamm.x_average(a - b) == pybamm.x_average(a) - pybamm.x_average(b)
 
-    def test_size_average(self):
+    @given(st.integers(min_value=-(2**63), max_value=2**63 - 1))
```
they are used to made sure that the random values generated are strictly a signed 64-bit integer .

prady0t replied:
@kratman Here's an example of how hypothesis could be helpful. Any value out of this range throws an error. Even though this is a very large value and does not make any sense, we may have such discrepancies with other functions.

RohitP2005 commented on tests/unit/test_expression_tree/test_averages.py:
```diff
@@ -267,8 +271,9 @@ def test_r_average(self):
 
-    def test_yz_average(self):
-        a = pybamm.Scalar(1)
+    @given(st.integers(min_value=-(2**63), max_value=2**63 - 1))
```
explained here at https://github.com/pybamm-team/PyBaMM/pull/4724#discussion_r1900290087

RohitP2005 commented:
> I've added reviews for `test_averages.py`. I'll add reviews for `test_unary_operators.py` once the failing tests are fixed.

I've updated the heading as requested, and committed all your suggested changes except for one.
Regarding the implementation of Hypothesis, I limited its use to specific tests to avoid adding overhead by applying it to too many cases. Do you think the way I implemented Hypothesis is optimized, or are there areas where I can improve further?

prady0t commented on tests/unit/test_expression_tree/test_unary_operators.py:
```diff
@@ -65,15 +69,17 @@ def test_negation(self, mocker):
 
-    def test_absolute(self, mocker):
+    @settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
+    @given(random_value=st.floats(min_value=-1e3, max_value=1e3))
```
```suggestion
    @given(st.floats(min_value=-1e38, max_value=0))
```
Since we are only working with negative numbers.

prady0t commented on tests/unit/test_expression_tree/test_unary_operators.py:
```diff
@@ -65,15 +69,17 @@ def test_negation(self, mocker):
         assert absa.children[0].name == a.name
 
-        b = pybamm.Scalar(-4)
+        b = pybamm.Scalar(-random_value)
```
```suggestion
        b = pybamm.Scalar(random_value)
```

prady0t commented on tests/unit/test_expression_tree/test_unary_operators.py:
```diff
@@ -65,15 +69,17 @@ def test_negation(self, mocker):
+        b = pybamm.Scalar(-random_value)
         absb = pybamm.AbsoluteValue(b)
-        assert absb.evaluate() == 4
+        assert absb.evaluate() == abs(random_value)
```
```suggestion
        assert absb.evaluate() == -random_value
```

prady0t commented on tests/unit/test_expression_tree/test_unary_operators.py:
```diff
@@ -153,19 +165,24 @@ def test_sign(self):
-    def test_floor(self, mocker):
+    @settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
+    @given(
+        random_value1=st.floats(min_value=-1e3, max_value=1e3),
+        random_value2=st.floats(min_value=-1e3, max_value=1e3),
+    )
```
```suggestion
        random_value1=st.floats(min_value=-1e38, max_value=1e38),
        random_value2=st.floats(min_value=-1e38, max_value=1e38),
    )
```
If there's a particular reason you are keeping for small range here and everywhere else, I'll be glad to know.

prady0t commented on tests/unit/test_expression_tree/test_unary_operators.py:
```diff
@@ -181,19 +198,24 @@ def test_floor(self, mocker):
 
-    def test_ceiling(self, mocker):
+    @settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
+    @given(
+        random_value1=st.floats(min_value=-1e3, max_value=1e3),
+        random_value2=st.floats(min_value=-1e3, max_value=1e3),
```
```suggestion
    @given(
        random_value1=st.floats(min_value=-1e38, max_value=1e38),
        random_value2=st.floats(min_value=-1e38, max_value=1e38),
```

prady0t commented on tests/unit/test_expression_tree/test_unary_operators.py:
```diff
@@ -116,17 +122,23 @@ def test_smooth_absolute_value(self):
-    def test_sign(self):
-        b = pybamm.Scalar(-4)
+    @given(
+        random_value=st.integers(),
+        random_matrix=st.lists(
+            st.floats(min_value=-10, max_value=10), min_size=5, max_size=5
+        ),
```
Can we increase the size here? i'll look into this bit later

RohitP2005 replied:
okay , i guess 10 will be good number

RohitP2005 replied:
hey @prady0t shall i go with ```min_size=10``` also i thought of creating a new PR by closing this one, cause it sflooded with too much of commits

agriyakhetarpal replied:
I think this PR is fine; it preserves the review comments in one place. Once the PR and the reviews are completed, please feel free to squash unnecessary commits as an exercise, or we can do it as we squash-merge PRs anyway.

prady0t commented on tests/unit/test_expression_tree/test_unary_operators.py:
```diff
@@ -25,15 +27,17 @@ def test_unary_operator(self):
 
-    def test_negation(self, mocker):
+    @settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
+    @given(random_value=st.floats(min_value=-1e3, max_value=1e3))
```
```suggestion
    @given(random_value=st.floats(min_value=-1e38, max_value=1e38))
```

kratman commented:
@RohitP2005 you keep merging develop into this branch, but there are still a bunch of failing tests. Can you fix those so the review can continue

RohitP2005 commented:
> @RohitP2005 you keep merging develop into this branch, but there are still a bunch of failing tests. Can you fix those so the review can continue

Sorry about that, my previous commit passed the CI test, after the new branch merge it is failing.

I will look into it asap.

RohitP2005 commented:
Hey @agriyakhetarpal  can u trigger the CI workflows for this branch

kratman commented:
@RohitP2005 Triggered CI and updated the branch

Saransh-cpp commented:
Requesting a review from @prady0t on this.

RohitP2005 commented:
> @RohitP2005 Triggered CI and updated the branch

thanks @kratman 
i will look into the failing docs build asap

RohitP2005 commented:
Hey @agriyakhetarpal  I need updates on this PR. Its been there for a long time.
Looking forward to complete this and move on to the next one.

agriyakhetarpal commented on pyproject.toml:
```diff
@@ -41,6 +41,7 @@ dependencies = [
     "posthog",
     "pyyaml",
     "platformdirs",
+    "hypothesis",
```
Could you please move this to the "dev" extras below in the file?

RohitP2005 replied:
@agriyakhetarpal Took care of that also with the latest commit

agriyakhetarpal reviewed (commented):
Done. Sorry for the delay in the review, @RohitP2005! Thanks for your work here. I just have one comment – happy to approve after that.

agriyakhetarpal commented:
Hi @prady0t, I plan to merge this early next week so that @RohitP2005 can proceed further – unless you have comments by then.

RohitP2005 commented:
@agriyakhetarpal I would like a final check by @prady0t 
before merging this.
Cause there's lot other tests that need property based testing.
It would helpful if you suggest what to do next on this matter.

kratman commented:
> Cause there's lot other tests that need property based testing. It would helpful if you suggest what to do next on this matter.

I think I would like to see this be limited to only where it is needed and useful rather than all tests. We would not want to cause the complexity and runtime of our tests to explode from excessive testing. The functions you have done so far are fairly straight forward, but tests in other parts of the code can be quite slow and complicated.

My other concern is that the codebase is physics based, so you need to make sure that the value ranges make sense for the codebase as well. Some parameters may only be valid in the 0-1 range or need to be only positive floats.

prady0t commented:
Other than the above review, everything look good. Thanks @RohitP2005
